src/ATen/native/xpu/sycl/LossCTCKernels.cpp

```diff
@@ -0,0 +1,1266 @@
+  if (targets.dim() == 1) { // concatenated targets
+    int64_t pos = 0;
+    for (int64_t i = 0; i < batch_size; i++) {
+      TORCH_CHECK(
+          target_lengths[i] >= 0,
+          "Expected target_lengths to have value at least ",
+          0,
+          ", but got value ",
+          target_lengths[i],
+          " (while checking arguments for ",
+          c,
+          ")");
+      tg_batch_offsets_data[i] = pos;
+      pos += target_lengths[i];
+      if (max_target_length < target_lengths[i])
+        max_target_length = target_lengths[i];
+    }
+    tg_target_stride = targets.stride(0);
+    checkSize(c, targets_arg, 0, pos);
+  } else { // batch x max_target_length
+    // dim is 2
+    int64_t tg_batch_stride = targets.stride(0);
+    for (int64_t i = 0; i < batch_size; i++) {
+      TORCH_CHECK(
+          target_lengths[i] >= 0,
+          "Expected target_lengths to have value at least ",
+          0,
+          ", but got value ",
+          target_lengths[i],
+          " (while checking arguments for ",
+          c,
+          ")");
+      tg_batch_offsets_data[i] = i * tg_batch_stride;
+      if (max_target_length < target_lengths[i])
+        max_target_length = target_lengths[i];
+    }
+    tg_target_stride = targets.stride(1);
+    checkSize(c, targets_arg, 0, batch_size);
+    TORCH_CHECK(
+        targets.size(1) >= max_target_length,
+        "Expected tensor to have size at least ",
+        max_target_length,
+        " at dimension 1, but got size ",
+        targets.size(1),
+        " for ",
+        targets_arg,
+        " (while checking arguments for ",
+        c,
+        ")");
+  }
+  int64_t max_input_length = log_probs.size(0);
+  for (int64_t b = 0; b < batch_size; b++) {
+    TORCH_CHECK(
+        input_lengths[b] >= 0,
+        "Expected input_lengths to have value at least ",
+        0,
+        ", but got value ",
+        input_lengths[b],
+        " (while checking arguments for ",
+        c,
+        ")");
+    TORCH_CHECK(
+        input_lengths[b] <= max_input_length,
+        "Expected input_lengths to have value at most ",
+        max_input_length,
+        ", but got value ",
+        input_lengths[b],
+        " (while checking arguments for ",
+        c,
+        ")");
+  }
+
+  auto target_lengths_t =
+      at::tensor(target_lengths, targets.options().dtype(kLong));
+  auto input_lengths_t =
+      at::tensor(input_lengths, targets.options().dtype(kLong));
+  tg_batch_offsets = tg_batch_offsets.to(targets.device());
+
+  Tensor log_alpha = at::empty(
+      {batch_size, log_probs.size(0), 2 * max_target_length + 1},
+      log_probs.options());
+  Tensor neg_log_likelihood = at::empty({batch_size}, log_probs.options());
+
+  // Very likely, we could be more clever here, e.g. learning (or generalizing
+  // and reusing) from SoftMax.cu...
+  constexpr int max_threads = std::is_same<scalar_t, float>::value
+      ? 1024
```

Comment:
Which is not device compatible, if simd16 and LGRF are enabled on PVC or SIMD8 is enabled on MTL, please use runtime query.